src/psyclone/domain/lfric/lfric_arg_descriptor.py

```diff
@@ -537,21 +569,38 @@ def _init_scalar(self, arg_type):
         # Check whether something other than a scalar is passed in
         if self._argument_type not in LFRicArgDescriptor.VALID_SCALAR_NAMES:
             raise InternalError(
-                "LFRicArgDescriptor._init_scalar(): expecting a scalar "
+                "LFRicArgDescriptor._init_scalar(): Expected a scalar "
                 "argument but got an argument of type '{0}'.".
                 format(arg_type.args[0]))
 
-        # There must be 2 arguments to describe a scalar
-        if self._nargs != 2:
+        # There must be 3 argument descriptors to describe a scalar.
+        # TODO in #874: Remove support for the old-style 2 descriptors.
+        min_scalar_nargs = 2 + self._offset
+        if self._nargs != min_scalar_nargs:
             raise ParseError(
-                "In the LFRic API each 'meta_arg' entry must have 2 "
+                "In the LFRic API each 'meta_arg' entry must have {0} "
                 "arguments if its first argument is 'gh_{{r,i}}scalar', but "
-                "found {0} in '{1}'.".format(self._nargs, arg_type))
-
-        # Scalar data_type is determined by its metadata descriptor for
-        # type as a first argument, however this will change to the second
-        # argument in issue #774
-        self._data_type = arg_type.args[0].name
+                "found {1} in '{2}'.".
+                format(min_scalar_nargs, self._nargs, arg_type))
+
+        # Check whether an invalid data type for a scalar argument is passed
+        # in. Valid data types for scalars are valid data types in LFRic API.
+        # TODO in #874: Remove the support for old-style scalar metadata that
+        #               assigns the data type from the scalar name (the 1st
+        #               argument).
+        #               Note: The main check for the valid scalar data types
+        #               will be ParseError in the class constructor and this
+        #               scalar init method only needs to check for
+        #               InternalError.
+        if not self._data_type and self._offset == 0:
+            self._data_type = arg_type.args[0].name
+        if (self._data_type not in
+                LFRicArgDescriptor.VALID_SCALAR_DATA_TYPES):
+            raise InternalError(
```

Comment:
I added the `ParseError` in the constructor, as well as the relevant test for it. As said above, this will need to be modified when data type is introduced for fields and operators.